packages/block-library/src/table-of-contents/hooks.js

```diff
@@ -65,15 +65,7 @@ function getLatestHeadings( select, clientId ) {
 
 	/** The page (of a paginated post) a heading will be part of. */
 	let headingPage = 1;
-	let headingPageLink = null;
-
-	// If the core/editor store is available, we can add permalinks to the
-	// generated table of contents.
-	if ( typeof permalink === 'string' ) {
-		headingPageLink = isPaginated
-			? addQueryArgs( permalink, { page: headingPage } )
-			: permalink;
-	}
```

Comment:
Removing this logic will break the ToC block in paginated pages/posts. If I remember correctly, that was the reason it requires a permalink.